### PR TITLE
refactor(typings): many cool changes

### LINKED ISSUE #5015: Move ChannelType typings to within discord.js module
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Eg. I'm always frustrated when [...]

The `ChannelType` enum typings are declared at the top of `typings/index.d.ts`, outside of the `discord.js` module. Thus, they appear at the root of all TypeScript projects which have discord.js in their packages. 

Type Definition:
https://github.com/discordjs/discord.js/blob/b8fd3f65d964f080afb42dd37eb21a13c0d84f8f/typings/index.d.ts#L1-L10 

For example, if I import `discord.js` as follows:

```
import * as Discord from "discord.js";
```

then I would expect to access `ChannelType` as such:

```
let myType = Discord.ChannelType.text;
```

Instead, `ChannelType` appears at the root:

```
let myType = ChannelType.text;
```

and gives an error at runtime:

```
ReferenceError: ChannelType is not defined
```

**Describe the ideal solution**

Move `ChannelType` typings into the `discord.js` module.

**Describe alternatives you've considered**

I need to use this for `GuildCreateChannelOptions` as seen here:
https://github.com/discordjs/discord.js/blob/b8fd3f65d964f080afb42dd37eb21a13c0d84f8f/typings/index.d.ts#L2562-L2565

So for now I just use the `any` type and assign `"text"` or `"voice"` as needed.

**Additional context**

I've noticed a near duplicate inside `Constants`:

https://github.com/discordjs/discord.js/blob/b8fd3f65d964f080afb42dd37eb21a13c0d84f8f/typings/index.d.ts#L556-L564

Not sure what to do with this.

### PR DESCRIPTION
**Pull request by vaporox**
**Please describe the changes this PR makes and why it should be merged:**

- Remove `declare module` to allow for npm aliases
- Remove `null` as a possible value for `Constants.UserAgent` as of #5113
- Replace `PathLike` with `string`
- Replace `{ [key: X]: Y }` with `Record<X, Y>`
​
- Move `AllowedImageFormat` from region Classes to region Typedefs
- Move `Collection` from region Collections to region Classes
- Move `version` and `Constants` into their own region Constants
​
- Remove overloads for `Collection#flatMap` and `Collection#mapValues` that already exist in `BaseCollection`
- Change the return type of `Collection#toJSON` to be an array which depends on `V`
- Create a type `Flatten<T>` used for `Util#flatten` and `Collection#toJSON`
​
- Update `index.ts` accordingly

Resolves #5015.

**Status**

- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**

- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: typings/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,2226 @@
...
+}
+
+export interface ChannelPosition {
+  channel: ChannelResolvable;
+  position: number;
+}
+
+export type ChannelResolvable = Channel | Snowflake;
+
+export enum ChannelType {
```
Comment (kyranet) [inline typings/index.d.ts:2223]: `ChannelType` shouldn't be exported, it's a private type made for convenience. It also does not exist as a JSDoc but as the description of [`Channel#type`](https://discord.js.org/#/docs/main/stable/class/Channel?scrollTo=type), therefore, it shouldn't be exported.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: typings/index.d.ts (hunk 1 above)
Comment (Vaporox) [reply typings/index.d.ts:2223]: You cannot declare enums in declaration files without exporting them, so it's either convert to a type (could do that here) or actually create an export in the source (would probably need another `fix:` pr)

Comment (vladfrangu) [reply typings/index.d.ts:2223]: Can you share the error you get by doing `declare enum ChannelType`? That's also what compiled TS code does with enums (it does `export declare enum <name>`)?

Comment (Vaporox) [reply typings/index.d.ts:2223]: It doesn't give an error but you can still import the enum that way

File: typings/index.d.ts [hunk 2]
```diff
@@ -9,3260 +2231,1007 @@ declare enum ChannelType {
...
+}
 
-  type SpeakingString = 'SPEAKING' | 'SOUNDSHARE' | 'PRIORITY_SPEAKING';
+export interface RolePosition {
+  role: RoleResolvable;
+  position: number;
+}
 
-  type StreamType = 'unknown' | 'converted' | 'opus' | 'ogg/opus' | 'webm/opus';
+export type RoleResolvable = Role | string;
```
Comment (NotSugden) [inline typings/index.d.ts:3091]:
```suggestion
export type RoleResolvable = Role | Snowflake;
```

File: typings/index.d.ts [hunk 3]
```diff
@@ -9,3260 +2231,1007 @@ declare enum ChannelType {
...
+  | 'DEAFEN_MEMBERS'
+  | 'MOVE_MEMBERS'
+  | 'USE_VAD'
+  | 'CHANGE_NICKNAME'
+  | 'MANAGE_NICKNAMES'
+  | 'MANAGE_ROLES'
+  | 'MANAGE_WEBHOOKS'
+  | 'MANAGE_EMOJIS';
+
+export interface RecursiveArray<T> extends ReadonlyArray<T | RecursiveArray<T>> {}
```
Comment (NotSugden) [inline typings/index.d.ts:2904]:
```suggestion
export type RecursiveArray<T> = ReadonlyArray<T | RecursiveArray<T>>;
```

File: typings/index.d.ts [hunk 4]
```diff
@@ -9,3260 +2231,1007 @@ declare enum ChannelType {
...
+
+export type OverwriteResolvable = PermissionOverwrites | OverwriteData;
+
+export type OverwriteType = 'member' | 'role';
+
+export interface PermissionFlags extends Record<PermissionString, number> {}
+
+export interface PermissionObject extends Record<PermissionString, boolean> {}
+
+export interface PermissionOverwriteOption extends Partial<Record<PermissionString, boolean | null>> {}
```
Comment (NotSugden) [inline typings/index.d.ts:2867]:
```suggestion
export type PermissionFlags = Record<PermissionString, number>;

export type PermissionObject = Record<PermissionString, boolean>;

export type PermissionOverwriteOption = Partial<Record<PermissionString, boolean | null>>;
```

File: typings/index.d.ts [hunk 5]
```diff
@@ -9,3260 +2231,1007 @@ declare enum ChannelType {
...
+  split?: boolean | SplitOptions;
+  replyTo?: MessageResolvable;
+}
+
+export type MessageReactionResolvable = MessageReaction | Snowflake;
+
+export interface MessageReference {
+  channelID: string;
+  guildID: string;
+  messageID: string | null;
```
Comment (NotSugden) [inline typings/index.d.ts:2827]:
```suggestion
  channelID: Snowflake;
  guildID: Snowflake;
  messageID: Snowflake | null;
```

**Review by NotSugden — CHANGES_REQUESTED** [review]
this diff is a bit tricky to read, i believe this branch is a few commits behind too, so possibly needs a rebase

File: typings/index.d.ts [hunk 6]
```diff
@@ -1,4 +1,2226 @@
...
+
+export interface ChannelPosition {
+  channel: ChannelResolvable;
+  position: number;
+}
+
+export type ChannelResolvable = Channel | Snowflake;
+
+export enum ChannelType {
   text = 0,
```
Comment (Mesteery) [inline typings/index.d.ts:2224]: An enum starts by default at 0. So there is no need to explicitely indicate the index.

**Review by Mesteery — CHANGES_REQUESTED** [review]

File: typings/index.d.ts (hunk 6 above)
Comment (vaporox) [reply typings/index.d.ts:2224]: that's only valid for source files, not for declaration files.

**Comment by vaporox** [comment]
Closing this as there are too many conflicts that can't get resolved easily. Might pick it up again towards the end of the v12 lifecycle

**Pull request closed without merging**